src/string.cr

```diff
@@ -1394,9 +1394,9 @@ class String
           buffer[i] = if byte >= 0x80
                         byte
                       elsif i.zero?
-                        byte.unsafe_chr.upcase.ord.to_u8!
+                        'a'.ord <= byte <= 'z'.ord ? byte - 32 : byte 
```

Comment:
```suggestion
                        'a'.ord <= byte <= 'z'.ord ? byte - 32 : byte
```

Reply:
benchmark is as below:
```
$./crystal build upcase.cr --release
$./upcase
string.downcase == string.downcase_pr12389 # => true
string.upcase == string.upcase_pr12389 # => true
a.titleize == a.capitalize # => true
a.capitalize_pr12389 == a.capitalize # => true
          downcase   4.16M (240.53ns) (± 6.35%)  144B/op   1.25× slower
            upcase   4.22M (237.01ns) (± 6.23%)  144B/op   1.23× slower
  downcase_pr12389   4.00M (249.79ns) (± 5.33%)  144B/op   1.30× slower
    upcase_pr12389   5.19M (192.52ns) (± 5.12%)  144B/op        fastest
          titleize   2.31M (432.41ns) (± 5.98%)  144B/op   2.25× slower
        capitalize   4.33M (230.99ns) (± 4.02%)  144B/op   1.20× slower
capitalize_pr12389   2.94M (340.52ns) (± 4.59%)  144B/op   1.77× slower

$./crystal -v
Crystal 1.5.0 [994c70b10] (2022-07-06)

LLVM: 10.0.0
Default target: x86_64-unknown-linux-gnu

$cat /etc/os-release |head -2
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"

$cat upcase.cr 
require "benchmark"

string = "hEllO WoRlD!." * 10
a = "hEllOWoRlD!." * 10

Benchmark.ips do |x|

  p! string.downcase == string.downcase_pr12389
  p! string.upcase == string.upcase_pr12389
  p! a.titleize == a.capitalize
  p! a.capitalize_pr12389 == a.capitalize

  x.report("downcase") do
    string.downcase
  end
  x.report("upcase") do
    string.upcase
  end

  x.report("downcase_pr12389") do
    string.downcase_pr12389
  end
  x.report("upcase_pr12389") do
    string.upcase_pr12389
  end

  x.report("titleize") do
    a.titleize
  end
  x.report("capitalize") do
    a.capitalize
  end
  x.report("capitalize_pr12389") do
    a.capitalize_pr12389
  end

end

```

Behind the benchmark, I added new functions to ```src/src/string.cr``` as below:
```
1380   def upcase_pr12389(options : Unicode::CaseOptions = :none) : String
1381     return self if empty?
1382 
1383     if single_byte_optimizable? && (options.none? || options.ascii?)
1384       return String.new(bytesize) do |buffer|
1385         bytesize.times do |i|
1386           byte = to_unsafe[i]
1387           buffer[i] = 'a'.ord <= byte <= 'z'.ord ? byte - 32 : byte
1388         end
1389         {@bytesize, @length}
1390       end
1391     end
1392 
1393     String.build(bytesize) { |io| upcase io, options }
1394   end
1395 

1329   def downcase_pr12389(options : Unicode::CaseOptions = :none) : String
1330     return self if empty?
1331 
1332     if single_byte_optimizable? && (options.none? || options.ascii?)
1333       return String.new(bytesize) do |buffer|
1334         bytesize.times do |i|
1335           byte = to_unsafe[i]
1336           buffer[i] = 'A'.ord <= byte <= 'Z'.ord ? byte + 32 : byte
1337         end
1338         {@bytesize, @length}
1339       end
1340     end
1341 
1342     String.build(bytesize) { |io| downcase io, options }
1343   end

1440   def capitalize_pr12389(options : Unicode::CaseOptions = :none) : String
1441     return self if empty?
1442 
1443     if single_byte_optimizable? && (options.none? || options.ascii?)
1444       return String.new(bytesize) do |buffer|
1445         bytesize.times do |i|
1446           byte = to_unsafe[i]
1447 
1448           buffer[i] = if byte >= 0x80
1449                         byte
1450                       elsif i.zero?
1451                         'a'.ord <= byte <= 'z'.ord ? byte - 32 : byte
1452                       else
1453                         byte.unsafe_chr.downcase.ord.to_u8!
1454                       end
1455         end
1456         {@bytesize, @length}
1457       end
1458     end
1459 
1460     String.build(bytesize) { |io| capitalize io, options }
1461   end

```

Reply:
the ```upcase``` and ```downcase``` of my above benchmark is different from https://github.com/crystal-lang/crystal/pull/12389 , I miss something?

Reply:
I update the code of ```capitalize_pr12389 `` and benchmark result as above~

Reply:
Here's my benchmark:

```crystal
require "benchmark"

a = "hEllOWoRlD!." * 10

Benchmark.ips do |x|
  x.report("capitalize") do
    a.capitalize
  end
  x.report("capitalize2") do
    a.capitalize2
  end
end
```

Here's my diff:

```crystal
diff --git a/src/string.cr b/src/string.cr
index e9dd8ab19..da4b1d01f 100644
--- a/src/string.cr
+++ b/src/string.cr
@@ -1406,6 +1406,29 @@ class String
     String.build(bytesize) { |io| capitalize io, options }
   end
 
+  def capitalize2(options : Unicode::CaseOptions = :none) : String
+    return self if empty?
+
+    if single_byte_optimizable? && (options.none? || options.ascii?)
+      return String.new(bytesize) do |buffer|
+        bytesize.times do |i|
+          byte = to_unsafe[i]
+
+          buffer[i] = if byte >= 0x80
+                        byte
+                      elsif i.zero?
+                        'a'.ord <= byte <= 'z'.ord ? byte - 32 : byte
+                      else
+                        'A'.ord <= byte <= 'Z'.ord ? byte + 32 : byte
+                      end
+        end
+        {@bytesize, @length}
+      end
+    end
+
+    String.build(bytesize) { |io| capitalize io, options }
+  end
+
   # Writes a capitalized version of `self` to the given *io*.
   #
   # ```
```

Here are the results I get:

```
 capitalize  28.74M ( 34.79ns) (± 3.15%)  144B/op   1.20× slower
capitalize2  34.37M ( 29.09ns) (± 3.48%)  144B/op        fastest
```

Can you try the same on your end?

Reply:
ok, I download ```crystal-1.5.0-1-linux-x86_64-bundled.tar.gz``` for crystal github release page( download the latest source code will cost too much time because of low download speed from github), and unzip to modify string.cr as below(copy from yours):
```
$diff  ./crystal-1.5.0-1/share/crystal/src/string.cr ./crystal-1.5.0-1/share/crystal/src/string.cr.raw 
1409,1431d1408
<   def capitalize2(options : Unicode::CaseOptions = :none) : String
<     return self if empty?
< 
<     if single_byte_optimizable? && (options.none? || options.ascii?)
<       return String.new(bytesize) do |buffer|
<         bytesize.times do |i|
<           byte = to_unsafe[i]
< 
<           buffer[i] = if byte >= 0x80
<                         byte
<                       elsif i.zero?
<                         'a'.ord <= byte <= 'z'.ord ? byte - 32 : byte
<                       else
<                         'A'.ord <= byte <= 'Z'.ord ? byte + 32 : byte
<                       end
<         end
<         {@bytesize, @length}
<       end
<     end
< 
<     String.build(bytesize) { |io| capitalize io, options }
<   end
< 

```

then use same benchmark code as yours to ```crystal build --release```, then get this:
```
 capitalize   4.99M (200.51ns) (± 0.84%)  144B/op        fastest
capitalize2   3.62M (276.45ns) (± 1.24%)  144B/op   1.38× slower
```
my cpu is ```Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz```.

Reply:
@orangeSi What operating system do you use?

Reply:
```
$cat /etc/os-release |head -2
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"

$./crystal -v
Crystal 1.5.0 [994c70b10] (2022-07-06)

LLVM: 10.0.0
Default target: x86_64-unknown-linux-gnu
```

Reply:
I think this will break in turkish locale. In that locale `"i".upcase` is not `"I"`.

Reply:
I thought this was guarded by `options.none? || options.ascii?`, but I'm not so sure anymore. Why is `Unicode::CaseOptions` a flag enum?

Reply:
> Why is Unicode::CaseOptions a flag enum?

Hm, maybe a mistake!